File: FBSimulatorControl/Commands/FBSimulatorLocationCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <CoreSimulator/SimDevice.h>
#import "FBSimulatorLocationCommands.h"

#import "FBSimulator.h"

@interface FBSimulatorLocationCommands ()

@property (nonatomic, weak, readonly) FBSimulator *simulator;

@end

@implementation FBSimulatorLocationCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBSimulator *)targets
{
  return [[self alloc] initWithSimulator:targets];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;

  return self;
}

#pragma mark FBSimulatorLocationCommands Protocol Implementation

- (FBFuture<NSNull *> *)overrideLocationWithLongitude:(double)longitude latitude:(double)latitude
{
  return [FBFuture onQueue:self.simulator.workQueue resolveValue:^NSNull *(NSError **error) {
    if (![self.simulator.device setLocationWithLatitude:latitude andLongitude:longitude error:error]) {
      return nil;
    }
    return NSNull.null;
  }];
}

@end

File: FBSimulatorControl/Commands/FBSimulatorLocationCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
import FBControlCore
import Foundation

public final class FBSimulatorLocationCommands {

  // MARK: - Properties

  private weak var simulator: FBSimulator?

  // MARK: - Initializers

  public class func commands(with simulator: FBSimulator) -> FBSimulatorLocationCommands {
    FBSimulatorLocationCommands(simulator: simulator)
  }

  private init(simulator: FBSimulator) {
    self.simulator = simulator
  }

  // MARK: - Private

  fileprivate func overrideLocationAsync(longitude: Double, latitude: Double) async throws {
    guard let simulator = self.simulator else {
      throw FBWeakTargetError.simulator
    }
    try simulator.device.setLocationWithLatitude(latitude, andLongitude: longitude)
  }
}

// MARK: - FBSimulator+LocationCommands

extension FBSimulator: LocationCommands {

  public func overrideLocation(longitude: Double, latitude: Double) async throws {
    try await locationCommands().overrideLocationAsync(longitude: longitude, latitude: latitude)
  }
}
